File: php/src/Advertisement/Domain/AdvertisementRepository.php
<?php
declare(strict_types=1);

namespace Demo\App\Advertisement\Domain;

use Chemem\Bingo\Functional\Functors\Monads\Either;
use Demo\App\Advertisement\Domain\Model\Advertisement;
use Demo\App\Advertisement\Domain\Model\ValueObject\AdvertisementId;
use Demo\App\Common\Result;

interface AdvertisementRepository
{
    public function save(Advertisement $advertisement): Either;

    public function findById(AdvertisementId $id): Either;
}

File: kotlin/src/main/kotlin/advertisement/domain/AdvertisementRepository.kt
package advertisement.domain

import advertisement.domain.model.Advertisement
import advertisement.domain.model.value_object.AdvertisementId
import arrow.core.Either

interface AdvertisementRepository {
    fun save(advertisement: Advertisement):  Either<Any, Advertisement>
    fun findById(id: AdvertisementId): Either<Any, Advertisement>
}